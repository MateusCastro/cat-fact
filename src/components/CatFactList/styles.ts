import styled from "styled-components";

export const Table = styled.table`
  border-collapse: collapse;
  border: 1px solid #e4e4e4;

  th {
    padding: 8px;
    text-align: left;
    background-color: #303030;
    color: white;
  }

  td {
    padding: 8px;
  }

  tr {
    background-color: #fff;
  }

  tr:nth-child(even) {
    background-color: #ececec;
  }
`;

export const Loading = styled.h2`
  text-align: center;
`;
