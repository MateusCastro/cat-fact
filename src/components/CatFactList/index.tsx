import React, { useEffect } from 'react';

import { useSelector, useDispatch } from 'react-redux';
import { ApplicationState } from '../../store';

import { Table, Loading } from './styles';

import { CatFact } from '../../common/types';
import { loadRequest } from '../../store/modules/catFacts/actions';
import CatFactItem from '../CatFactItem';

const CatFactList: React.FC = () => {
  const dispatch = useDispatch();
  const catFacts = useSelector<ApplicationState, CatFact[]>(
    (state) => state.catFacts.data,
  );
  const loading = useSelector<ApplicationState, boolean>(
    (state) => state.catFacts.loading,
  );

  useEffect(() => {
    dispatch(loadRequest());
  }, [dispatch]);

  return (
    <>
      {loading ? (
        <Loading>Loading...</Loading>
      ) : (
        <Table data-testid="carFact-table">
          <thead>
            <tr>
              <th>addded by</th>
              <th>Fact</th>
              <th>Votes</th>
            </tr>
          </thead>
          <tbody>
            {catFacts.map((catFact) => (
              <CatFactItem catFact={catFact} key={catFact._id} />
            ))}
          </tbody>
        </Table>
      )}
    </>
  );
};

export default CatFactList;
