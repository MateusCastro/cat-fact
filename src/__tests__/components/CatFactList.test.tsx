import React from 'react';
import { mount, ReactWrapper } from 'enzyme';

import { useSelector, useDispatch } from 'react-redux';

import CatFactList from '../../components/CatFactList';

jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
  useDispatch: jest.fn(),
}));

const mockUseSelector = useSelector as jest.Mock;
const mockUseDispatch = useDispatch as jest.Mock;
const mockDispatch = jest.fn();

describe('CarFactList component', () => {
  let wrapper: ReactWrapper;

  beforeEach(() => {
    mockUseDispatch.mockImplementation(() => mockDispatch);
    jest.spyOn(React, 'useEffect').mockImplementation((f) => f());

    mockUseSelector.mockImplementation((cb) =>
      cb({
        catFacts: {
          data: [
            {
              _id: 'asdf23s2d1f',
              text: 'gato loves grace',
              upvotes: 8,
              user: null,
            },
            {
              _id: '96512312sf',
              text: 'Cats sleep 70% of their lives.',
              upvotes: 7,
              user: {
                name: {
                  first: 'Mateus',
                  last: 'Domingos',
                },
              },
            },
          ],
        },
      }),
    );

    wrapper = mount(<CatFactList />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should load the CatFact list', () => {
    expect(mockDispatch).toHaveBeenCalled();
  });

  it('should render CarFact list', () => {
    expect(wrapper.find('tbody').find('CatFactItem')).toHaveLength(2);
  });

  it('should render loading', () => {
    mockUseSelector.mockImplementation((cb) =>
      cb({
        catFacts: {
          loading: true,
        },
      }),
    );
    wrapper = mount(<CatFactList />);
    expect(wrapper.find('h2').text()).toEqual('Loading...');
    expect(wrapper).toMatchSnapshot();
  });
});
