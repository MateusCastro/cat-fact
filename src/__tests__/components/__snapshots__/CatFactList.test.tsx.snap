// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarFactList component should load the CatFact list 1`] = `
<CatFactList>
  <styled.table
    data-testid="carFact-table"
  >
    <table
      className="sc-AxjAm tEjqf"
      data-testid="carFact-table"
    >
      <thead>
        <tr>
          <th>
            addded by
          </th>
          <th>
            Fact
          </th>
          <th>
            Votes
          </th>
        </tr>
      </thead>
      <tbody>
        <CatFactItem
          catFact={
            Object {
              "_id": "asdf23s2d1f",
              "text": "gato loves grace",
              "upvotes": 8,
              "user": null,
            }
          }
          key="asdf23s2d1f"
        >
          <tr>
            <td>
              Unknown
            </td>
            <td>
              gato loves grace
            </td>
            <td>
              8
            </td>
          </tr>
        </CatFactItem>
        <CatFactItem
          catFact={
            Object {
              "_id": "96512312sf",
              "text": "Cats sleep 70% of their lives.",
              "upvotes": 7,
              "user": Object {
                "name": Object {
                  "first": "Mateus",
                  "last": "Domingos",
                },
              },
            }
          }
          key="96512312sf"
        >
          <tr>
            <td>
              Mateus Domingos
            </td>
            <td>
              Cats sleep 70% of their lives.
            </td>
            <td>
              7
            </td>
          </tr>
        </CatFactItem>
      </tbody>
    </table>
  </styled.table>
</CatFactList>
`;

exports[`CarFactList component should render CarFact list 1`] = `
<CatFactList>
  <styled.table
    data-testid="carFact-table"
  >
    <table
      className="sc-AxjAm tEjqf"
      data-testid="carFact-table"
    >
      <thead>
        <tr>
          <th>
            addded by
          </th>
          <th>
            Fact
          </th>
          <th>
            Votes
          </th>
        </tr>
      </thead>
      <tbody>
        <CatFactItem
          catFact={
            Object {
              "_id": "asdf23s2d1f",
              "text": "gato loves grace",
              "upvotes": 8,
              "user": null,
            }
          }
          key="asdf23s2d1f"
        >
          <tr>
            <td>
              Unknown
            </td>
            <td>
              gato loves grace
            </td>
            <td>
              8
            </td>
          </tr>
        </CatFactItem>
        <CatFactItem
          catFact={
            Object {
              "_id": "96512312sf",
              "text": "Cats sleep 70% of their lives.",
              "upvotes": 7,
              "user": Object {
                "name": Object {
                  "first": "Mateus",
                  "last": "Domingos",
                },
              },
            }
          }
          key="96512312sf"
        >
          <tr>
            <td>
              Mateus Domingos
            </td>
            <td>
              Cats sleep 70% of their lives.
            </td>
            <td>
              7
            </td>
          </tr>
        </CatFactItem>
      </tbody>
    </table>
  </styled.table>
</CatFactList>
`;

exports[`CarFactList component should render loading 1`] = `
<CatFactList>
  <styled.table
    data-testid="carFact-table"
  >
    <table
      className="sc-AxjAm tEjqf"
      data-testid="carFact-table"
    >
      <thead>
        <tr>
          <th>
            addded by
          </th>
          <th>
            Fact
          </th>
          <th>
            Votes
          </th>
        </tr>
      </thead>
      <tbody>
        <CatFactItem
          catFact={
            Object {
              "_id": "asdf23s2d1f",
              "text": "gato loves grace",
              "upvotes": 8,
              "user": null,
            }
          }
          key="asdf23s2d1f"
        >
          <tr>
            <td>
              Unknown
            </td>
            <td>
              gato loves grace
            </td>
            <td>
              8
            </td>
          </tr>
        </CatFactItem>
        <CatFactItem
          catFact={
            Object {
              "_id": "96512312sf",
              "text": "Cats sleep 70% of their lives.",
              "upvotes": 7,
              "user": Object {
                "name": Object {
                  "first": "Mateus",
                  "last": "Domingos",
                },
              },
            }
          }
          key="96512312sf"
        >
          <tr>
            <td>
              Mateus Domingos
            </td>
            <td>
              Cats sleep 70% of their lives.
            </td>
            <td>
              7
            </td>
          </tr>
        </CatFactItem>
      </tbody>
    </table>
  </styled.table>
</CatFactList>
`;

exports[`CarFactList component should render loading 2`] = `
<CatFactList>
  <styled.h2>
    <h2
      className="sc-AxirZ cqPXHd"
    >
      Loading...
    </h2>
  </styled.h2>
</CatFactList>
`;
